File: JULY-2025/14-07-2025/p1290.ex
# Definition for singly-linked list.
#
# defmodule ListNode do
#   @type t :: %__MODULE__{
#           val: integer,
#           next: ListNode.t() | nil
#         }
#   defstruct val: 0, next: nil
# end

defmodule Solution do
  @spec get_decimal_value(head :: ListNode.t | nil) :: integer
  def get_decimal_value(head), do: do_get_decimal_value(head, 0)

  defp do_get_decimal_value(%ListNode{val: v, next: next}, acc), do: do_get_decimal_value(next, (acc * 2) + v)
  defp do_get_decimal_value(_, acc), do: acc
end

File: JULY-2025/14-07-2025/p1290.erl
%% Definition for singly-linked list.
%%
%% -record(list_node, {val = 0 :: integer(),
%%                     next = null :: 'null' | #list_node{}}).

gdv(L,N) ->
    case L#list_node.next == null of
        true -> N + L#list_node.val;
        false -> gdv(L#list_node.next, 
            (N + L#list_node.val) bsl 1)
    end
.

-spec get_decimal_value(Head :: #list_node{} | null) -> integer().
get_decimal_value(Head) ->
    gdv(Head,0)
.